### PR TITLE
New callout style for curved lines

### PR DESCRIPTION
**Pull request by nyalldawson**
Renders a nice cartographically pleasing curved line between the
labels and features.

Options include selecting a specific curve orientation (clockwise
or counterclockwise), or an automatic orientation option which
determines optimal orientation for each individual label. Users
also have control over the amount of curvature applied to the
callout lines.

Here's a quick screencast of the automatic orientation option:

Peek 2021-03-18 16-04

### CODE REVIEW

**Review by nirvn — APPROVED** [review]
Beautiful. One missing icon (a different PR is fine) and one question :)

File: src/core/geometry/qgsgeometryutils.cpp [hunk 1]
```diff
@@ -1385,6 +1402,83 @@ QStringList QgsGeometryUtils::wktGetChildBlocks( const QString &wkt, const QStri
   return blocks;
 }
 
+int QgsGeometryUtils::closestSideOfRectangle( double right, double bottom, double left, double top, double x, double y )
```
[the remark this answers is not in the document]
Comment (nyalldawson) [reply src/core/geometry/qgsgeometryutils.cpp:1405]: This class is intended to be low level, highly optimised for a range of uses. And in this case we have QPointF/QRectF objects originally, so I want to skip the unnecessary cost of conversion to the Qgis classes...

[the remark this answers is not in the document]
Comment (nirvn) [reply src/core/geometry/qgsgeometryutils.cpp:1405]: OK -- wondering if it'd still be useful to have a sibling function that'd take a QgsRectangle and QgsPointXY (merely to forward it to the more optimized function) as an additional API offering.

I've got no strong feelings here really, just thinking out-loud.

**Comment by nyalldawson** [comment]
@nirvn I've added icon requests to https://github.com/qgis/QGIS/projects/1

**Pull request merged**